Do amortized clean up in WeakHashSet

cdumez opened the pull request:
#### 06391ab46e0212df1e08aeb8965eda7d8e1d9d99
<pre>
Do amortized clean up in WeakHashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=247712">https://bugs.webkit.org/show_bug.cgi?id=247712</a>

Reviewed by Ryosuke Niwa.

Do amortized clean up in WeakHashSet, similarly to what we already do in
WeakHashMap.

* Source/WTF/wtf/WeakHashSet.h:
* Source/WebCore/dom/SlotAssignment.cpp:

Canonical link: <a href="https://commits.webkit.org/256541@main">https://commits.webkit.org/256541@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f175915345458a39aa37f0a921a86411b87b4911

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96076 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5325 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29122 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105628 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165956 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100054 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5442 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34099 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88455 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101445 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101736 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82682 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31064 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87782 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87085 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39817 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82452 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37492 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20654 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28244 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4523 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42088 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85131 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43978 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39918 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19213 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -155,7 +172,28 @@ class WeakHashSet final {
+        m_operationCountSinceLastCleanup = 0;
+    }
+
+    ALWAYS_INLINE unsigned increaseOperationCountSinceLastCleanup(unsigned operationsPerformed = 1) const
```
Looks like this argument is always 1 so unnecessary??

rniwa commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -155,7 +172,28 @@ class WeakHashSet final {
+        return currentCount;
+    }
+
+    ALWAYS_INLINE void amortizedCleanupIfNeeded(unsigned operationsPerformed = 1) const
```
Ditto.

rniwa commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -155,7 +172,28 @@ class WeakHashSet final {
+    }
+
     WeakPtrImplSet m_set;
+    mutable unsigned m_operationCountSinceLastCleanup { 0 };
```
Hm... it's not great to increase the size of this object.

cdumez committed:
Do amortized clean up in WeakHashSet
https://bugs.webkit.org/show_bug.cgi?id=247712

Reviewed by Ryosuke Niwa.

Do amortized clean up in WeakHashSet, similarly to what we already do in
WeakHashMap.

* Source/WTF/wtf/WeakHashSet.h:
* Source/WebCore/dom/SlotAssignment.cpp:

Canonical link: https://commits.webkit.org/256541@main
